### PR TITLE
Add credential checking and encoding to the abstract service

### PR DESCRIPTION
**Pull request by dejanb**

### CODE REVIEW

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java [hunk 1]
```diff
@@ -44,6 +65,11 @@ public void setTenantInformationService(final TenantInformationService tenantInf
         this.tenantInformationService = tenantInformationService;
     }
 
+    @Autowired
+    public void setPasswordEncoder(final HonoPasswordEncoder passwordEncoder) {
+        this.passwordEncoder = passwordEncoder;
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java]: does it make sense to allow setting this to `null`?

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java [hunk 2]
```diff
@@ -91,4 +122,51 @@ public void setTenantInformationService(final TenantInformationService tenantInf
                         : processReadCredentials(DeviceKey.from(result.getPayload(), deviceId), span));
     }
 
+    private Future<List<CommonCredential>> verifyAndEncodePasswords(final List<CommonCredential> credentials) {
+        // Check if we need to encode passwords
+        if (!needToEncode(credentials)) {
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java]: personally, I find it easier to read (and understand) the *positive* condition, i.e.
```
if (needToEncode(credentials)) {
} else {
}
```

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java [hunk 3]
```diff
@@ -91,4 +122,51 @@ public void setTenantInformationService(final TenantInformationService tenantInf
...
+     * @return {@code true} is the list contains at least one password which needs to be encoded on the
+     *         server side.
+     */
+    private static boolean needToEncode(final List<CommonCredential> credentials) {
+        return credentials
+                .stream()
+                .filter(PasswordCredential.class::isInstance)
+                .map(PasswordCredential.class::cast)
+                .flatMap(c -> c.getSecrets().stream())
+                .anyMatch(s -> s.getPasswordPlain() != null && !s.getPasswordPlain().isEmpty());
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java]: `!Strings.isNullOrEmpty(s.getPasswordPlain())`?

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java (hunk 1 above)
Comment (dejanb) [reply services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java]: I added `requireNonNull` check for tenant information service as well. We can introduce Noop version for folks that don't want to use autowired one.

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java [hunk 4]
```diff
@@ -91,4 +131,51 @@ public void setTenantInformationService(final TenantInformationService tenantInf
                         : processReadCredentials(DeviceKey.from(result.getPayload(), deviceId), span));
     }
 
+    private Future<List<CommonCredential>> verifyAndEncodePasswords(final List<CommonCredential> credentials) {
+        // Check if we need to encode passwords
+        if (!needToEncode(credentials)) {
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java]: now you also need to remove the `!` ;-)

Comment (dejanb) [reply services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/credentials/AbstractCredentialsManagementService.java]: Gah! thanks :)

**Pull request merged**